### PR TITLE
Provides the core API for the Virtual File System as Layer3 of the st…

### PR DESCRIPTION
**Pull request by andyHa**
…orage framework.

### CODE REVIEW

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 1]
```diff
@@ -0,0 +1,794 @@
...
+    }
+
+    /**
+     * Determines if the file represents a directory.
+     *
+     * @return <tt>true</tt> if the file represents a directory, <tt>false</tt> otherwise
+     */
+    public boolean isDirectory() {
+        try {
+            if (directoryFlagSupplier == null) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: I think inverting this if check would improve readability/flow

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 2]
```diff
@@ -0,0 +1,794 @@
...
+    }
+
+    /**
+     * Tries to delete this file.
+     *
+     * @return <tt>true</tt> if the operation was successful, <tt>false</tt> otherwise
+     */
+    public boolean tryDelete() {
+        try {
+            if (deleteHandler == null) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: I think inverting this if check would improve readability/flow and woudl be consistent to the other methods in this class

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 3]
```diff
@@ -0,0 +1,794 @@
...
+    }
+
+    /**
+     * Tries to rename this file.
+     *
+     * @return <tt>true</tt> if the operation was successful, <tt>false</tt> otherwise
+     */
+    public boolean tryRename(String newName) {
+        try {
+            if (renameHandler == null) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: I think inverting this if check would improve readability/flow and woudl be consistent to the other methods in this class

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 4]
```diff
@@ -0,0 +1,794 @@
...
+    /**
+     * Tries to efficiently move the file in the given directory.
+     *
+     * @param newParent the new parent directory
+     * @return <tt>true</tt> if the operation was successful, <tt>false</tt> otherwise
+     */
+    public boolean tryFastMoveTo(VirtualFile newParent) {
+        try {
+            if (moveHandler == null) {
+                return false;
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: I think inverting this if check would improve readability/flow and woudl be consistent to the other methods in this class

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 5]
```diff
@@ -0,0 +1,794 @@
...
+     * @throws HandledException if the file cannot be efficiently moved
+     */
+    public void fastMoveTo(VirtualFile newParent) {
+        if (!tryFastMoveTo(newParent)) {
+            throw Exceptions.createHandled().withNLSKey("VirtualFile.cannotMove").set("file", path()).handle();
+        }
+    }
+
+    /**
+     * Determines if this file can bei moved (either via an efficient implementation or by copy+delete).
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]:
```suggestion
     * Determines if this file can be moved (either via an efficient implementation or by copy+delete).
```

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 6]
```diff
@@ -0,0 +1,794 @@
...
+
+    /**
+     * Determines if the file exists.
+     *
+     * @return <tt>true</tt> if the file exists, <tt>false</tt> otherwise
+     */
+    public boolean exists() {
+        try {
+            if (existsFlagSupplier == null) {
+                return false;
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: I think inverting this if check would improve readability/flow and would be consistent to the other methods in this class

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 7]
```diff
@@ -0,0 +1,794 @@
...
+     * Tries to find a child with the given name.
+     * <p>
+     * Note that the given file can be non-existent.
+     *
+     * @param name the name of the child to find
+     * @return the child wrapped as optional or an empty optional if the given name cannot be resolved into a file.
+     */
+    public Optional<VirtualFile> findChild(String name) {
+        try {
+            if (childProvider == null) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: I think inverting this if check would improve readability/flow and woudl be consistent to the other methods in this class

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 8]
```diff
@@ -0,0 +1,794 @@
...
+    /**
+     * Tries to resolve the relative path within this directory.
+     *
+     * @param relativePath the path to resolve
+     * @return the relative path wrapped as optional or an  empty optional if the given relaive path cannot be
+     * resolved into a file.
+     */
+    public Optional<VirtualFile> tryResolve(String relativePath) {
+        if (relativePath != null && relativePath.startsWith("/")) {
+            relativePath = relativePath.substring(1);
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: Extract method maybe? Altering a parameter is a code smell (doesn't SonarQube warn about this? It did before)

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 9]
```diff
@@ -0,0 +1,794 @@
...
+        }
+
+        if (Strings.isEmpty(relativePath)) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: Could be moved to the top of this function

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 10]
```diff
@@ -0,0 +1,794 @@
...
+    }
+
+    /**
+     * Determines if children can be created for this file.
+     *
+     * @return <tt>true</tt> if children can (probably) be created, <tt>false</tt> otherwise
+     */
+    public boolean canCreateChildren() {
+        try {
+            if (canCreateChildrenHandler == null) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: I think inverting this if check would improve readability/flow and would be consistent to the other methods in this class

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 11]
```diff
@@ -0,0 +1,794 @@
...
+    }
+
+    /**
+     * Tries to create an output stream to write to the file.
+     *
+     * @return an output stream to provide the contents of the child or an empty optional if no output stream can be created
+     */
+    public Optional<OutputStream> tryCreateOutputStream() {
+        try {
+            if (outputStreamSupplier == null) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: I think inverting this if check would improve readability/flow and would be consistent to the other methods in this class

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 12]
```diff
@@ -0,0 +1,794 @@
...
+    }
+
+    /**
+     * Tries to create an input stream to read the contents of the file.
+     *
+     * @return an input stream to read the contents of the child or an empty optional if no input stream can be created
+     */
+    public Optional<InputStream> tryCreateInputStream() {
+        try {
+            if (inputStreamSupplier == null) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: I think inverting this if check would improve readability/flow and would be consistent to the other methods in this class

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 13]
```diff
@@ -0,0 +1,794 @@
...
+                                                                  .handle());
+    }
+
+    @Override
+    public String toString() {
+        return path();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: obj -> other?

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 14]
```diff
@@ -0,0 +1,794 @@
...
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
+        return path().hashCode();
+    }
+
+    @Override
+    public int compareTo(VirtualFile o) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: o -> other?

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 15]
```diff
@@ -0,0 +1,794 @@
...
+
+        return Objects.compare(path(), o.path(), String::compareTo);
+    }
+
+    /**
+     * Delivers the contents of this file into the given web context.
+     *
+     * @param ctx the HTTP request to respond to
+     */
+    public void deliverTo(WebContext ctx) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: ctx -> webContext?

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 16]
```diff
@@ -0,0 +1,794 @@
...
+               .error(HttpResponseStatus.INTERNAL_SERVER_ERROR, handleErrorInCallback(e, "tunnelHandler"));
+        }
+    }
+
+    /**
+     * Delivers the contents of this file as download.
+     *
+     * @param ctx the HTTP request to respond to
+     */
+    public void deliverDownloadTo(WebContext ctx) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: ctx -> webContext?

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 17]
```diff
@@ -0,0 +1,794 @@
...
+            tunnelHandler.accept(this, ctx.respondWith().named(name()));
+        } catch (Exception e) {
+            ctx.respondWith()
+               .error(HttpResponseStatus.INTERNAL_SERVER_ERROR, handleErrorInCallback(e, "tunnelHandler"));
+        }
+    }
+
+    @SuppressWarnings("UnstableApiUsage")
+    private static void defaultTunnelHandler(VirtualFile file, Response response) {
+        try (InputStream in = file.createInputStream()) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: in -> input?

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 18]
```diff
@@ -0,0 +1,794 @@
...
+            if (in == null) {
+                response.error(HttpResponseStatus.NOT_FOUND);
+                return;
+            }
+
+            try (OutputStream out = response.outputStream(HttpResponseStatus.OK,
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: out -> output?

File: src/main/java/sirius/biz/storage/layer3/VirtualFileSystem.java [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+ * Provides the main entrypoint into the <b>Virtual File System</b>.
+ * <p>
+ * This serves mainly three purposes:
+ * <ol>
+ * <li>Making internally stored files accessible to the outside by using <b>downlinks</b> like FTP, SFTP or SCP</li>
+ * <li>Making external files accessible to the software by using <b>uplinks</b> like FS, CIFS, SFTP</li>
+ * <li>Provide a uniform management UI for all file based operations within sirius</li>
+ * </ol>
+ */
+@Register(classes = VirtualFileSystem.class)
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFileSystem.java]: No framework annotation?

File: src/main/java/sirius/biz/storage/layer3/VirtualFileSystem.java [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+public class VirtualFileSystem {
+
+    private VirtualFile root;
+
+    @PriorityParts(VFSRoot.class)
+    private List<VFSRoot> rootProviders;
+
+    private class RootProvider implements ChildProvider {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFileSystem.java]: JavaDoc even if its an inner class?

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java (hunk 9 above)
Comment (andyHa) [reply src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: nope substring(1) modifies...

File: src/main/java/sirius/biz/storage/layer3/VirtualFileSystem.java (hunk 19 above)
Comment (andyHa) [reply src/main/java/sirius/biz/storage/layer3/VirtualFileSystem.java]: does no harm - therefore there is no need to disable...

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java (hunk 9 above)
Comment (sabieber) [reply src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: Oh right, forgot the case where the path equals "/"

**Pull request merged**